File: samples/python/43.complex-dialog/dialogs/main_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs import (
    ComponentDialog,
    WaterfallDialog,
    WaterfallStepContext,
    DialogTurnResult,
)
from botbuilder.core import MessageFactory, UserState

from data_models import UserProfile
from dialogs.top_level_dialog import TopLevelDialog


class MainDialog(ComponentDialog):
    def __init__(self, user_state: UserState):
        super(MainDialog, self).__init__(MainDialog.__name__)

        self.user_state = user_state

        self.add_dialog(TopLevelDialog(TopLevelDialog.__name__))
        self.add_dialog(
            WaterfallDialog("WFDialog", [self.initial_step, self.final_step])
        )

        self.initial_dialog_id = "WFDialog"

    async def initial_step(
        self, step_context: WaterfallStepContext
    ) -> DialogTurnResult:
        return await step_context.begin_dialog(TopLevelDialog.__name__)

    async def final_step(self, step_context: WaterfallStepContext) -> DialogTurnResult:
        user_info: UserProfile = step_context.result

        companies = (
            "no companies"
            if len(user_info.companies_to_review) == 0
            else " and ".join(user_info.companies_to_review)
        )
        status = f"You are signed up to review {companies}."

        await step_context.context.send_activity(MessageFactory.text(status))

        # store the UserProfile
        accessor = self.user_state.create_property("UserProfile")
        await accessor.set(step_context.context, user_info)

        return await step_context.end_dialog()

File: samples/java_springboot/43.complex-dialog/src/main/java/com/microsoft/bot/sample/complexdialog/MainDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.complexdialog;

import com.microsoft.bot.builder.StatePropertyAccessor;
import com.microsoft.bot.builder.UserState;
import com.microsoft.bot.dialogs.ComponentDialog;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.dialogs.WaterfallDialog;
import com.microsoft.bot.dialogs.WaterfallStep;
import com.microsoft.bot.dialogs.WaterfallStepContext;
import java.util.Arrays;
import java.util.concurrent.CompletableFuture;

public class MainDialog extends ComponentDialog {
    private UserState userState;

    public MainDialog(UserState withUserState) {
        super("MainDialog");

        userState = withUserState;

        addDialog(new TopLevelDialog("TopLevelDialog"));

        WaterfallStep[] waterfallSteps = {
            this::initialStep,
            this::finalStep
        };

        addDialog(new WaterfallDialog("WaterfallDialog", Arrays.asList(waterfallSteps)));

        setInitialDialogId("WaterfallDialog");
    }

    private CompletableFuture<DialogTurnResult> initialStep(WaterfallStepContext stepContext) {
        return stepContext.beginDialog("TopLevelDialog");
    }

    private CompletableFuture<DialogTurnResult> finalStep(WaterfallStepContext stepContext) {
        UserProfile userInfo = (UserProfile) stepContext.getResult();

        String status = String.format("You are signed up to review %s.",
            userInfo.getCompaniesToReview().size() == 0
                ? "no companies"
                : String.join(" and ", userInfo.getCompaniesToReview()));

        return stepContext.getContext().sendActivity(status)
            .thenCompose(resourceResponse -> {
                StatePropertyAccessor<UserProfile> userProfileAccessor = userState.createProperty("UserProfile");
                return userProfileAccessor.set(stepContext.getContext(), userInfo);
            })
            .thenCompose(setResult -> stepContext.endDialog());
    }
}
